nova/virt/vmwareapi/vmops.py

```diff
@@ -3102,3 +3183,64 @@ def _get_vm_monitor_spec(self, vim):
             [property_spec, property_spec_vm],
             [object_spec])
         return property_filter_spec
+
+    def place_vm(self, context, instance):
+        # We currently only fill the bare-minimum to get a placement.
+        # The datastore for the VM is selected as on instance creation,
+        # instead of allowing placevm to decide it (which may be better)
+        # We also do not pass the information about the NICs and other
+        # attached disks, which may give the placement a better information
+        client_factory = self._session.vim.client.factory
+        flavor = instance.flavor
+        image_meta = instance.image_meta
+        image_info = images.VMwareImage.from_image(context,
+                                                   instance.image_ref,
+                                                   image_meta)
+
+        extra_specs = self._get_extra_specs(flavor, image_meta)
+
+        vi = self._get_vm_config_info(instance, image_info, extra_specs)
+
+        # Sets cpuAllocation, memoryAllocation, numCPUs, memoryMB
+        config_spec = vm_util.get_vm_resize_spec(client_factory,
+                                                    int(flavor.vcpus),
+                                                    int(flavor.memory_mb),
+                                                    extra_specs)
+        # The last mandatory field for config_spec per doc
+        config_spec.version = extra_specs.hw_version
+
+        vm_folder = self._get_project_folder(
+                vi.dc_info, project_id=instance.project_id, type_='Instances')
+
+        relocate_spec = vm_util.relocate_vm_spec(
+            client_factory,
+            res_pool=self._root_resource_pool,
+            datastore=vi.datastore.ref,
+            disk_move_type="moveAllDiskBackingsAndDisallowSharing",
+            folder=vm_folder)
+
+        placement_spec = client_factory.create("ns0:PlacementSpec")
+        placement_spec.placementType = "relocate"
+        # Maybe we want to allow that? Default is True
+        # placement_spec.disallowPrerequisiteMoves = False
+        placement_spec.configSpec = config_spec
+        placement_spec.relocateSpec = relocate_spec
+
+        # So we do not place the vm on failover host
+        placement_spec.hosts, _ = \
+            vm_util.get_hosts_and_reservations_for_cluster(
+                self._session, self._cluster)
+
+        return self._session._call_method(self._session.vim, "PlaceVm",
+            self._cluster, placementSpec=placement_spec)
```

Comment:
This might fail if DRS is disabled. Maybe we could check in the pre-live-migration check if DRS is not disabled?

Reply:
I would suggest to table that for a later change. I agree, it isn't functional complete (doesn't work without DRS),
but I rather than take the placement strategy from the upstream driver (or my own I used before making PlaceVM work).